---
title: "HS.ST.prepandmerge"
author: "Harry Smith"
date: "5/2/2018"
output: html_document
---

###################################################
# Project: HS Panel
# Author: Harry Smith
# Synopsis: The purpose of this script is to generate
# File lists BY brain region of the HS gtfs from
# stringtie for use in the stringtie merge step
####################################################


```{r, echo=FALSE, message=FALSE, warning=FALSE}
## set up workspace
rm(list=ls())
options(stringsAsFactors = FALSE)

dirs <- "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/"

## Read in lists
batch1 <- read.table(file = paste(dirs, "file.list.2106-04-05.txt", sep = ""), 
                     header = FALSE, sep = "\t")
batch2 <- read.table(file = paste(dirs, "file.list.2106-10-10.txt", sep = ""), 
                     header = FALSE, sep = "\t")
batch345 <- read.table(file = paste(dirs, "filelist.batch_3_4_5.txt", sep = ""), 
                      header = FALSE, sep = "\t")

## only keep files ending in .gtf
b1.tmp <- batch1$V1[endsWith(batch1$V1, ".gtf")]
b2.tmp <- batch2$V1[endsWith(batch2$V1, ".gtf")]
b345.tmp <- batch345$V1[endsWith(batch345$V1, ".gtf")]
b.list <- list("b1" = b1.tmp, "b2" = b2.tmp, "b345" = b345.tmp)

## Get unique brain regions in each batch
b1.br <- unique(unlist(lapply(strsplit(b1.tmp[-length(b1.tmp)], split = "_", fixed = TRUE), FUN = function(x) x[2])))
b2.br <- unique(unlist(lapply(strsplit(b2.tmp[-length(b2.tmp)], split = "_", fixed = TRUE), FUN = function(x) x[2])))
b345.br <- unique(unlist(lapply(strsplit(b345.tmp[-length(b345.tmp)], split = "[_.]+", fixed = FALSE), FUN = function(x) x[2])))

## Create brain region specific file lists
b.out <- list()
for(j in names(b.list)){
  b.out.tmp <- list()
  for(i in 1:5) {
    b.out.tmp[[i]] <- b.list[[j]][grepl(b1.br[i], b.list[[j]])]
  }
  b.out[[j]] <- b.out.tmp
}

b1.dir <- "/data/hi-seq/HS.Brain.polyA.2016-04-05/reconstruction/"
b2.dir <- "/data/hi-seq/HS.Brain.polyA.2016-10-10/reconstruction/"
b345.dir <- "/data/hi-seq/HS.Brain.polyA.batch3_4_5/reconstruction"
b.out[[1]] <- lapply(b.out[[1]], FUN = function(x) paste(b1.dir, x, sep = ""))
b.out[[2]] <- lapply(b.out[[2]], FUN = function(x) paste(b2.dir, x, sep = ""))
b.out[[3]] <- lapply(b.out[[3]], FUN = function(x) paste(b345.dir, x, sep = ""))

brcat <- list()
for(i in 1:5) {
  h1 <- b.out[[1]][[i]]
  h2 <- b.out[[2]][[i]]
  h3 <- b.out[[3]][[i]]
  hcat <- c(h1, h2, h3)
  brcat[[b1.br[i]]] <- hcat
}

## check number of files in each list
lapply(brcat, FUN = function(x) {length(x)})

## add file path to each list

## write file lists
for(i in 1:5) {
  write.table(brcat[[i]], file = paste(dirs, names(brcat)[i], "List.txt", sep = ""), row.names = FALSE, col.names = FALSE, quote = FALSE)
}

## create file list for all samples
filesFull <- unlist(brcat)
write.table(filesFull, file = paste(dirs, "filesFull", "List.txt", sep = ""), row.names = FALSE, col.names = FALSE, quote = FALSE)
```



```
## Acbc
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_Acbc.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/AcbcList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_Acbc.gtf

## IL
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_IL.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/ILList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_IL.gtf

## LHB
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_LHB.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/LHBList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_LHB.gtf

## PL
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_PL.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/PLList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_PL.gtf

## VoLo
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_VoLo.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/VoLoList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_VoLo.gtf

## FULL
stringtie --merge -m 0 -T 0 -F 0 -p 8 -G ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf -o ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_FULL.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/fileLists/filesFullList.txt

/usr/local/gffcompare.0.10.1/gffcompare -G -r ~/Saba_Lab/RNA-Seq/HRDP/Transcript_Reconstruction/index/Rattus_norvegicus.Rnor_6.0.90.cleaned.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/stringtie_results_FULL.gtf
```

Generate and save gtf dataframe abjects for each brain region and the total dataset

```{r, echo=FALSE, message=FALSE, warning=FALSE}
## Set directory
directory <- "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/"
## Set each file name object
Acbc <- "gffcmp.annotated_Acbc.gtf"
IL <- "gffcmp.annotated_IL.gtf"
LHB <- "gffcmp.annotated_LHB.gtf"
PL <- "gffcmp.annotated_PL.gtf"
VoLo <- "gffcmp.annotated_VoLo.gtf"
FULL <- "gffcmp.annotated_FULL.gtf"

options(stringsAsFactors = FALSE)
library(rtracklayer) 
library(GenomicRanges)

Acbc.gtf <- import(paste(directory, Acbc, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
Acbc_df <- as.data.frame(Acbc.gtf) #Converts the object above to a dataframe 

IL.gtf <- import(paste(directory, IL, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
IL_df <- as.data.frame(IL.gtf) #Converts the object above to a dataframe 

LHB.gtf <- import(paste(directory, LHB, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
LHB_df <- as.data.frame(LHB.gtf) #Converts the object above to a dataframe 

PL.gtf <- import(paste(directory, PL, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
PL_df <- as.data.frame(PL.gtf) #Converts the object above to a dataframe 

VoLo.gtf <- import(paste(directory, VoLo, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
VoLo_df <- as.data.frame(VoLo.gtf) #Converts the object above to a dataframe 

FULL.gtf <- import(paste(directory, FULL, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
FULL_df <- as.data.frame(FULL.gtf) #Converts the object above to a dataframe 
save(Acbc_df, IL_df, LHB_df, PL_df, VoLo_df, FULL_df, file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/Brain/", "quantitation/", "HS.gtfdataframes.RData", sep = ""))

## remove ensembl transcripts from IL, PL, and LHB gtfs and write back out to gtf format for gffcompare
IL.out <- "gffcmp.annotated_IL_MSTRG.gtf"
PL.out <- "gffcmp.annotated_PL_MSTRG.gtf"
LHB.out <- "gffcmp.annotated_LHB_MSTRG.gtf"

IL_df.ss <- IL_df[grep("MSTRG", IL_df$transcript_id), ]
table(substr(IL_df.ss$transcript_id, start = 1, stop = 5)) # should only be MSTRG
export(IL_df.ss, paste(directory, IL.out, sep = ""))

PL_df.ss <- PL_df[grep("MSTRG", PL_df$transcript_id), ]
table(substr(PL_df.ss$transcript_id, start = 1, stop = 5)) # should only be MSTRG
export(PL_df.ss, paste(directory, PL.out, sep = ""))

LHB_df.ss <- LHB_df[grep("MSTRG", LHB_df$transcript_id), ]
table(substr(LHB_df.ss$transcript_id, start = 1, stop = 5)) # should only be MSTRG
export(LHB_df.ss, paste(directory, LHB.out, sep = ""))

```

##use gffcompare to create comparison gtfs for the following comparisons
1. IL vs. PL
2. IL vs. LHB

```
/usr/local/gffcompare.0.10.1/gffcompare -r ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/gffcmp.annotated_IL_MSTRG.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/gffcmp.annotated_PL_MSTRG.gtf

/usr/local/gffcompare.0.10.1/gffcompare -r ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/gffcmp.annotated_IL_MSTRG.gtf ~/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/gffcmp.annotated_LHB_MSTRG.gtf
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
## Set directory
directory <- "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/"
## Set each file name object
ILvPL <- "gffcmp.annotated_ILvPL_MSTRG.gtf"
ILvLHB <- "gffcmp.annotated_ILvLHB_MSTRG.gtf"


options(stringsAsFactors = FALSE)
library(rtracklayer) 
library(GenomicRanges)

ILvPL.gtf <- import(paste(directory, ILvPL, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
ILvPL_df <- as.data.frame(ILvPL.gtf) #Converts the object above to a dataframe 

ILvLHB.gtf <- import(paste(directory, ILvLHB, sep = "")) #This imports the .gtf file and formats the information in the file and the 9th column to something more    readible
ILvLHB_df <- as.data.frame(ILvLHB.gtf) #Converts the object above to a dataframe 
save(ILvPL_df, ILvLHB_df, file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/Transcriptome_Reconstruction/Brain/", "quantitation/", "HS.brainregiongtfcomparison.RData", sep = ""))
```


Generate transcript only (remove exon rows) data, and merge with .tmap file 

```{r, echo=FALSE, message=FALSE, warning=FALSE}
## Set directory where .tmap files can be found
dirs2 <- "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/stringtie/mergedgtfs/"

## Set each file name object
Acbc.t <- "gffcmp.stringtie_results_Acbc.gtf"
IL.t <- "gffcmp.stringtie_results_IL.gtf"
LHB.t <- "gffcmp.stringtie_results_LHB.gtf"
PL.t <- "gffcmp.stringtie_results_PL.gtf"
VoLo.t <- "gffcmp.stringtie_results_VoLo.gtf"
FULL.t <- "gffcmp.stringtie_results_FULL.gtf"

Acbc.tmap <- read.table(file = paste(dirs2, Acbc.t, ".tmap", sep = ""), header = TRUE)
IL.tmap <- read.table(file = paste(dirs2, IL.t, ".tmap", sep = ""), header = TRUE)
LHB.tmap <- read.table(file = paste(dirs2, LHB.t, ".tmap", sep = ""), header = TRUE)
PL.tmap <- read.table(file = paste(dirs2, PL.t, ".tmap", sep = ""), header = TRUE)
VoLo.tmap <- read.table(file = paste(dirs2, VoLo.t, ".tmap", sep = ""), header = TRUE)
FULL.tmap <- read.table(file = paste(dirs2, FULL.t, ".tmap", sep = ""), header = TRUE)

## Remove exon rows from gtf datasets
sm <- function(data, tmap){
  a.tmp <- data[!duplicated(data$transcript_id), ]
  a.tlev <- merge(a.tmp, tmap, by.x = "transcript_id", by.y = "qry_id")
}

Acbc.tlev <- sm(Acbc_df, Acbc.tmap)
IL.tlev <- sm(IL_df, IL.tmap)
LHB.tlev <- sm(LHB_df, LHB.tmap)
PL.tlev <- sm(PL_df, PL.tmap)
VoLo.tlev <- sm(VoLo_df, VoLo.tmap)
FULL.tlev <- sm(FULL_df, FULL.tmap)

save(Acbc.tlev, IL.tlev, LHB.tlev, PL.tlev, VoLo.tlev, FULL.tlev, file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/Transcriptome_Reconstruction/Brain/", "quantitation/", "HS.gtf.tmap.merge.RData", sep = ""))
```
