---
title: "HS.ST.gtfsummarize"
author: "Harry Smith"
date: "5/11/2018"
output: html_document
---

This program reads in the gtf dataframes built in "HS.ST.prepandmerge.Rmd", and uses them to create a number of summary statistics and corresponding graphics.
1. number of transcripts/genes within each brain region transcriptome
2. number of novel transcripts within each brain region transcriptome
3. of the novel transcripts, how many are single exon and how many are multi exon
4. What gffcompare classcode is associated with each novel transcript

```{r, echo=FALSE, message=FALSE, warning=FALSE}
## set up workspace
rm(list=ls())
options(stringsAsFactors = FALSE)
## load datasets
load(file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/", "quantitation/", "HS.gtfdataframes.RData", sep = ""))
dflist <- list("Acbc" = Acbc_df, "IL" = IL_df, "LHB" = LHB_df, "PL" = PL_df, "VoLo" = VoLo_df, "Combined" = FULL_df)
ss <- function(data) {
  ## number of transcripts
  tcnt <- sum(!duplicated(data$transcript_id))
  ## number of genes
  gcnt <- sum(!duplicated(data$gene_id))
  ## number of novel transcripts
  nt <- data[grepl("MSTRG", data$transcript_id), ]
  ntcnt <- sum(!duplicated(nt$transcript_id))
  
  ## generate table of transcrpt IDs and their number of exons for novel transcript dataset
  nt$exon_number2 <- ifelse(is.na(nt$exon_number), 0, 1)
  nexcntdf <- aggregate(exon_number2~transcript_id, data = nt, function(x) sum(x))
  singleexcnt <- sum(nexcntdf$exon_number2 == 1)
  multiexcnt <- sum(nexcntdf$exon_number2 > 1)
  
  ## results vector
  out <- c(tcnt, ntcnt, multiexcnt, singleexcnt, gcnt)
  return(out)
}
forPlot <- as.data.frame(do.call("rbind", lapply(dflist, FUN = function(x) ss(x))))
forPlot$dataset <- rownames(forPlot)
forPlot$dataset <- factor(forPlot$dataset, levels = c("Acbc", "IL", "LHB",  "PL", "VoLo", "Combined"))
forPlot$dataset <- relevel(forPlot$dataset, 'Combined')
colnames(forPlot) <- c("Number of transcripts", "Number of novel transcripts", "Number of novel transcripts with multiple exons", "Number of novel transcripts with one exon", "Number of genes", "Dataset")

tpg <- function(x) {
  ## aggregate and summarize transcripts by gene
  ##Subset gtf to remove exons (i.e. transcript level). This number should match the totals in the forPlot table
  df <- x[!duplicated(x$transcript_id), ]
  ## Calculate the number of transcripts per gene
  ag <- aggregate(transcript_id~gene_id, data = df, FUN = function(x) sum(grepl(paste("^MSTRG", "ENSRNOT", sep = "|"), x)))
  ## Summarize transcripts per gene
  out <- format(round(summary(ag$transcript_id), 1), nsmall = 1)
  return(out)
}

tpgsummary <- as.data.frame(do.call("rbind", lapply(dflist, FUN = function(x) tpg(x))))
colnames(tpgsummary) <- c("Minimum", "1st Quartile", "Median", "Mean", "3rd Quartile", "Maximum")
write.csv(tpgsummary, file = "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/tables/HS.transcrptpergeneSummary.csv", row.names = T, quote = F)

## Generate summary table for combined dataset
dfFinal <- data.frame("Dataset" = forPlot$Dataset,
          "Number of transcripts" = prettyNum(forPlot$`Number of transcripts`, big.mark = ","),
          "Number of novel transcripts" = paste(prettyNum(forPlot$`Number of novel transcripts`, big.mark = ","), " ", "(", as.character(round((forPlot$`Number of novel transcripts`/forPlot$`Number of transcripts`)*100), 1), "%", ")", sep = ""),
          "Number of novel transcripts with multiple exons" = paste(prettyNum(forPlot$`Number of novel transcripts with multiple exons`, big.mark = ","), " ", "(", as.character(round((forPlot$`Number of novel transcripts with multiple exons`/forPlot$`Number of novel transcripts`)*100), 1), "%", ")", sep = ""),
          "Number of novel transcripts with one exon" = paste(prettyNum(forPlot$`Number of novel transcripts with one exon`, big.mark = ","), " ", "(", as.character(round((forPlot$`Number of novel transcripts with one exon`/forPlot$`Number of novel transcripts`)*100), 1), "%", ")", sep = ""), 
          "Number of genes" = prettyNum(forPlot$`Number of genes`, big.mark = ","))
colnames(dfFinal) <- c("Dataset", "Number of transcripts", "Number of novel transcripts", "Number of novel transcripts with multiple exons", "Number of novel transcripts with one exon", "Number of genes")
write.csv(dfFinal, file = "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/tables/HS.transcriptContentSummary.csv", row.names = F, quote = T)

```

Generate plots

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
library(reshape2)
library(RColorBrewer)
fp.melt <- melt(forPlot, id.vars = "Dataset")
png("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/graphics/HS.gtfTranscriptSummary.png", height = 6, width = 8, res = 480, units = "in")
ggplot(fp.melt, aes(Dataset, value)) +   
  geom_bar(aes(fill = variable), stat = 'identity', position = 'dodge') +
  labs(x = "Dataset (Brain region)", y = "Count") +
  scale_fill_manual(values=c("black", "darkorchid2", "dodgerblue3", "green4", "darkgoldenrod2"), 
                    name=NULL,
                       breaks=c("Number of transcripts", "Number of novel transcripts", "Number of novel transcripts with multiple exons", "Number of novel transcripts with one exon", "Number of genes"),
                       labels=c("Number of transcripts\n", "Number of novel transcripts\n", "Number of novel transcripts \nwith multiple exons\n", "Number of novel transcripts \nwith one exon\n", "Number of genes\n")) + ggtitle("Summary of transcript content \n by brain region")
dev.off()
```



```{r, echo=FALSE, message=FALSE, warning=FALSE}
## load gtf.tmap data
load(file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/", "quantitation/", "HS.gtf.tmap.merge.RData", sep = ""))
dflist2 <- list("Acbc" = Acbc.tlev, "IL" = IL.tlev, "LHB" = LHB.tlev, "PL" = PL.tlev, "VoLo" = VoLo.tlev, "Combined" = FULL.tlev)
## extract novel transcripts, and summarize class codes
ccs <- function(x){
  nt <- x[grepl("MSTRG", x$transcript_id), ]
  print(dim(nt)[1])
  out <- table(nt$class_code.x)
  return(out)
}
## Generate class code summary table
classcodesum <- as.data.frame(do.call("rbind", lapply(dflist2, FUN = function(x) ccs(x))))
classcodesum$Dataset <- rownames(classcodesum)
classcodesum$Dataset <- factor(classcodesum$Dataset, levels = c("Acbc", "IL", "LHB",  "PL", "VoLo", "Combined"))
classcodesum$Dataset <- relevel(classcodesum$Dataset, 'Combined')

## Generate class code summary table where o,p,s,x, and y are collapsed into one category
classcodeagg <- data.frame("Dataset" = classcodesum$Dataset, 
                           "j" = classcodesum$j, 
                           "u" = classcodesum$u, 
                           "i" = classcodesum$i, 
                           "other" = c(apply(classcodesum[, c("e", "o", "p", "s", "x", "y")], 1, FUN = function(x) sum(x))))

## Sanity check
identical(as.numeric(apply(classcodeagg[, -1], 1, function(x) sum(x))), as.numeric(forPlot$`Number of novel transcripts`))

## Generate class code summary table where the summary is the number transcripts in each class code as a percent of total novel transcripts
classcodepct <- data.frame("Dataset" = classcodeagg$Dataset, 
                           "j" = (classcodeagg$j/forPlot$`Number of novel transcripts`)*100, 
                           "u" = (classcodeagg$u/forPlot$`Number of novel transcripts`)*100, 
                           "i" = (classcodeagg$i/forPlot$`Number of novel transcripts`)*100, 
                           "other" = (classcodeagg$other/forPlot$`Number of novel transcripts`)*100)
## sanity check 
apply(classcodepct[, -1], 1, function(x) sum(x)) ## Should all be 100

## Melt datasets for plotting
cc.melt <- melt(classcodesum, id.vars = "Dataset")
ca.melt <- melt(classcodeagg, id.vars = "Dataset")
cp.melt <- melt(classcodepct, id.vars = "Dataset")

## Generate pretty table for class code summary
cc.pretty <- data.frame("Dataset" = classcodeagg$Dataset, 
                           "j" = paste(prettyNum(classcodeagg$j, big.mark = ","), " ", "(", as.character(round((classcodeagg$j/forPlot$`Number of novel transcripts`)*100)), "%)", sep = ""), 
                           "u" =  paste(prettyNum(classcodeagg$u, big.mark = ","), " ", "(", as.character(round((classcodeagg$u/forPlot$`Number of novel transcripts`)*100)), "%)", sep = ""), 
                           "i" =  paste(prettyNum(classcodeagg$i, big.mark = ","), " ", "(", as.character(round((classcodeagg$i/forPlot$`Number of novel transcripts`)*100)), "%)", sep = ""), 
                           "other" =  paste(prettyNum(classcodeagg$other, big.mark = ","), " ", "(", as.character(round((classcodeagg$other/forPlot$`Number of novel transcripts`)*100)), "%)", sep = ""))
colnames(cc.pretty) <- c("Dataset", "At least 1 exon junction match", "Intergenic", "Intronic", "Ambiguous overlap")
write.csv(cc.pretty, file = "/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/tables/HS.classcodeSummary.csv", row.names = F, quote = T)

## Generate plot summarizing all class codes
png("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/graphics/HS.gtfClassCodeSummary.png", height = 6, width = 8, res = 480, units = "in")
ggplot(cc.melt, aes(Dataset, value)) +   
  geom_bar(aes(fill = variable), stat = 'identity', position = 'dodge') +
  labs(x = "Dataset (Brain region)", y = "Count") +
  scale_fill_manual(values = brewer.pal(n = 9, name = "Set3"),
                    name="Class code",
                       breaks=c("e", "i", "j", "o", "p", "s", "u", "x", "y"),
                       labels=c("e", "i", "j", "o", "p", "s", "u", "x", "y")) + 
  ggtitle("Summary of class codes for \n novel transcripts") 
dev.off()

## Generate plot summarizing class codes with collapsed category
png("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/graphics/HS.gtfClassCodeCollapsedSummary.png", height = 6, width = 8, res = 480, units = "in")
ggplot(ca.melt, aes(Dataset, value)) +   
  geom_bar(aes(fill = variable), stat = 'identity', position = 'dodge') +
  labs(x = "Dataset (Brain region)", y = "Count") +
  scale_fill_manual(values = brewer.pal(n = 4, name = "Set3"),
                    name="Class code",
                       breaks=c("j", "u", "i", "other"),
                       labels=c("At least 1\n exon junction match\n", "Intergenic\n", "Intronic\n", "Ambiguous overlap\n")) + 
  ggtitle("Summary of class codes for \n novel transcripts") 
dev.off()

## Generate plot summarizing class codes with collapsed category and as a percent
png("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/graphics/HS.gtfClassCodePercentSummary.png", height = 6, width = 8, res = 480, units = "in")
ggplot(cp.melt, aes(Dataset, value)) +   
  geom_bar(aes(fill = variable), stat = 'identity', position = 'dodge') +
  labs(x = "Dataset (Brain region)", y = "Number of transcripts as a percent of \n total novel transcripts") +
  scale_fill_manual(values = brewer.pal(n = 4, name = "Set3"),
                    name="Class code",
                       breaks=c("j", "u", "i", "other"),
                       labels=c("At least 1\n exon junction match\n", "Intergenic\n", "Intronic\n", "Ambiguous overlap\n")) + 
  ggtitle("Summary of class codes as a \n percent of novel transcripts") 
dev.off()
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
## load gtf.tmap data
load(file = paste("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/", "quantitation/", "HS.brainregiongtfcomparison.RData", sep = ""))
ILvPL_df <- ILvPL_df[!duplicated(ILvPL_df$transcript_id), ]
ILvLHB_df <- ILvLHB_df[!duplicated(ILvLHB_df$transcript_id), ]
dflist3 <- list("ILvPL" = ILvPL_df, "ILvLHB" = ILvLHB_df)

## extract novel transcripts, and summarize class codes
ccs <- function(x){
  nt <- x[grepl("MSTRG", x$transcript_id), ]
  print(dim(nt)[1])
  out <- table(nt$class_code)
  return(out)
}
## Generate class code summary table
classcodecompare <- as.data.frame(do.call("rbind", lapply(dflist3, FUN = function(x) ccs(x))))
classcodecompare$Dataset <- rownames(classcodecompare)
classcodecompare$Dataset <- factor(classcodecompare$Dataset, levels = c("ILvPL", "ILvLHB"))

## Generate class code summary table where o,p,s,x, and y are collapsed into one category
classcodecompareagg <- data.frame("Dataset" = classcodecompare$Dataset,
                           "=" = classcodecompare$`=`, 
                           "j" = classcodecompare$j, 
                           "u" = classcodecompare$u, 
                           "i" = classcodecompare$i, 
                           "other" = c(apply(classcodecompare[, c("c", "e", "o", "p", "s", "x", "y")], 1, FUN = function(x) sum(x))))

identical(as.numeric(apply(classcodecompareagg[, -1], 1, function(x) sum(x))), as.numeric(c(dim(ILvPL_df)[1], dim(ILvLHB_df)[1])))

## Melt datasets for plotting
ccc.melt <- melt(classcodecompare, id.vars = "Dataset")
cca.melt <- melt(classcodecompareagg, id.vars = "Dataset")

## Generate plot summarizing class codes with collapsed category
png("/Volumes/smiharry/Saba_Lab/RNA-Seq/HS/graphics/HS.brainregiongtfcompare.png", height = 6, width = 8, res = 480, units = "in")
ggplot(ccc.melt, aes(Dataset, value)) +   
  geom_bar(aes(fill = variable), stat = 'identity', position = 'dodge') +
  labs(x = "Dataset (Brain region)", y = "Count") +
  scale_fill_manual(values = brewer.pal(n = 11, name = "Set3"),
                    name="Class code",
                       breaks=colnames(classcodecompare[-12]),
                       labels=colnames(classcodecompare[-12])) + 
  ggtitle("Class code summary after gffcompare where\n IL is the reference group") 
dev.off()

```